### PR TITLE
Feature 425 provide meta data for projects

### LINKED ISSUE #425: Provide meta data for projects
# Situation
Sometimes it is interesting to have additional meta data (having default keys and also custom keys) being readable by humans and machines as well.

**Example:** 
```
language.java=true
language.kotline=true
```

or 
```
acme.projectid=my-special-projectId
```

# Usecases
## Existing use cases
`UC_ADMIN_CREATES_PROJECT`
- optional set key values by REST when defined
- (+ provide this in DAUI with ENV variable containing default keys comma separated)

`UC_ADMIN_SHOWS_PROJECT_DETAILS`
- meta data is contained in project details as well

## New usecase
either we introduce 

`UC_ADMIN_UPDATES_PROJECT_METADATA`
"As an administrator, I want to update meta data key settings of my project (key, value) "
or 

`UC_ADMIN_UPDATES_PROJECT_DATA`
"As an administrator, I want to update description and meta data key settings of my project (key, value) "

in last case we could update description as well, but we do NOT change the owner, which is a own usecase and has more consequences (see #97 ). 

---
<sup>Albert Tregnaghi <albert.tregnaghi@daimler.com>, Daimler TSS GmbH, [imprint](https://github.com/Daimler/daimler-foss/blob/master/LEGAL_IMPRINT.md)</sup>

### LINKED ISSUE #464: WhiteListURI Validator not enforced on create and update project
At the moment the existing URIValidator is not enforced on create project and update project whitelist.

This should be enforced to answer wrong user requests with empty URI-Strings or exceeding 255 characters with 4xx-HTTP status instead of 500.

---
<sup>Albert Weißert <albert.weissert@daimler.com>, Daimler TSS GmbH, [imprint](https://github.com/Daimler/daimler-foss/blob/master/LEGAL_IMPRINT.md)</sup>

### PR DESCRIPTION
**Pull request by 4w31ss3**
closes #425 
closes #464

---
<sup>Albert Weißert <albert.weissert@daimler.com>, Daimler TSS GmbH, [imprint](https://github.com/Daimler/daimler-foss/blob/master/LEGAL_IMPRINT.md)</sup>

### CODE REVIEW

File: sechub-administration/src/test/java/com/daimler/sechub/domain/administration/project/ProjectUpdateAdministrationRestControllerMockTest.java [hunk 1]
```diff
@@ -104,6 +112,49 @@ public Void answer(InvocationOnMock invocation) {
...
+	
+	@Test
+	public void when_validator_marks_no_errors___calling_update_project_metadata_calls_update_service_and_returns_http_200() throws Exception {
+
+		/* execute + test @formatter:off */
+        this.mockMvc.perform(
+        		post(https(PORT_USED).buildUpdateProjectMetaData(projectId)).
+        		contentType(MediaType.APPLICATION_JSON_VALUE).
+        		content("{\"metaData\":{\"key1\":\"value1\",\"key2\":\"value2\"}}")
+        		)./*andDo(print()).*/
```
Comment (Jeeppler) [inline sechub-administration/src/test/java/com/daimler/sechub/domain/administration/project/ProjectUpdateAdministrationRestControllerMockTest.java:124]: Please remove the commented code.

File: sechub-doc/src/test/java/com/daimler/sechub/restdoc/ProjectUpdateAdministrationRestControllerRestDocTest.java [hunk 2]
```diff
@@ -95,6 +100,34 @@ public void restdoc_update_white_list_for_project() throws Exception {
...
+	@Test
+	@UseCaseRestDoc(useCase=UseCaseUpdateProjectMetaData.class)
+	public void restdoc_update_metadata_for_project() throws Exception {
+
+		/* execute + test @formatter:off */
+        this.mockMvc.perform(
+        		post(https(PORT_USED).buildUpdateProjectMetaData(PROJECT_ID.pathElement()),"projectId1").
+        		contentType(MediaType.APPLICATION_JSON_VALUE).
+        		content("{\"apiVersion\":\"1.0\", \"metaData\":{\"key1\":\"value1\"}}")
+        		)./*andDo(print()).*/
```
Comment (Jeeppler) [inline sechub-doc/src/test/java/com/daimler/sechub/restdoc/ProjectUpdateAdministrationRestControllerRestDocTest.java:113]: Please remove the unused code.

File: sechub-doc/src/test/java/com/daimler/sechub/restdoc/ProjectUpdateAdministrationRestControllerRestDocTest.java [hunk 3]
```diff
@@ -95,6 +100,34 @@ public void restdoc_update_white_list_for_project() throws Exception {
...
+        			andExpect(status().isOk()).
+        			andDo(document(RestDocPathFactory.createPath(UseCaseUpdateProjectMetaData.class),
+        					pathParameters(
```
Comment (Jeeppler) [inline sechub-doc/src/test/java/com/daimler/sechub/restdoc/ProjectUpdateAdministrationRestControllerRestDocTest.java:116]: Would it be possible to reduce the spacing a bit.

**Review by Jeeppler — CHANGES_REQUESTED** [review]
See comments.

File: sechub-administration/src/test/java/com/daimler/sechub/domain/administration/project/ProjectUpdateAdministrationRestControllerMockTest.java (hunk 1 above)
Comment (4w31ss3) [reply sechub-administration/src/test/java/com/daimler/sechub/domain/administration/project/ProjectUpdateAdministrationRestControllerMockTest.java:124]: Can do. I adopted this from already existing Tests in this and other classes and thought this was left on purpose to give a hint for fast debugging.

@de-jcup would it be ok for you if I remove this `andDo(print())` statements?

Comment (de-jcup) [reply sechub-administration/src/test/java/com/daimler/sechub/domain/administration/project/ProjectUpdateAdministrationRestControllerMockTest.java:124]: This was something I did when I started creating restdoc tests. So it is very easy to have communication printed out - suitable for creating and fixing rest doc tests. You just remove the comment parts and you got the output...

And I copied the comment... often ... and so did @4w31ss3 

I think we should either accept this one or remove all inside the rest doc tests.

File: sechub-doc/src/test/java/com/daimler/sechub/restdoc/ProjectUpdateAdministrationRestControllerRestDocTest.java (hunk 2 above)
Comment (de-jcup) [reply sechub-doc/src/test/java/com/daimler/sechub/restdoc/ProjectUpdateAdministrationRestControllerRestDocTest.java:113]: see my thoughts above - either keep or we create an issue to remove all.

File: sechub-administration/src/test/java/com/daimler/sechub/domain/administration/project/ProjectUpdateAdministrationRestControllerMockTest.java (hunk 1 above)
Comment (Jeeppler) [reply sechub-administration/src/test/java/com/daimler/sechub/domain/administration/project/ProjectUpdateAdministrationRestControllerMockTest.java:124]: In this case keep them.

File: sechub-administration/src/main/java/com/daimler/sechub/domain/administration/project/ProjectMetaData.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+import java.io.Serializable;
+
+public class ProjectMetaData implements Serializable {
+
+	private static final long serialVersionUID = -721445637539759118L;
+	
+	private String key;
+	private String value;
+	
+	public ProjectMetaData() {
```
Comment (de-jcup) [inline sechub-administration/src/main/java/com/daimler/sechub/domain/administration/project/ProjectMetaData.java:12]: If its only for JPA (and maybe for some unit testing) I would appreciate if you could make it package private like done at
`ScanAccess`.

File: sechub-administration/src/main/java/com/daimler/sechub/domain/administration/project/ProjectMetaDataEntry.java [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+	public static final String COLUMN_PROJECT_ID = "PROJECT_ID";
+	public static final String COLUMN_METADATA_KEY = "METADATA_KEY";
+	public static final String COLUMN_METADATA_VALUE = "METADATA_VALUE";
+	
+	/* +-----------------------------------------------------------------------+ */
+	/* +............................ JPQL .....................................+ */
+	/* +-----------------------------------------------------------------------+ */
+	public static final String CLASS_NAME = ProjectMetaDataEntry.class.getSimpleName();
+	
+	public ProjectMetaDataEntry() {
```
Comment (de-jcup) [inline sechub-administration/src/main/java/com/daimler/sechub/domain/administration/project/ProjectMetaDataEntry.java:31]: JPA only .. I would appreciate then a package private (same as my comment before)

File: sechub-administration/src/test/java/com/daimler/sechub/domain/administration/project/ProjectUpdateAdministrationRestControllerMockTest.java (hunk 1 above)
Comment (4w31ss3) [reply sechub-administration/src/test/java/com/daimler/sechub/domain/administration/project/ProjectUpdateAdministrationRestControllerMockTest.java:124]: Team decided to remove all /*andDo(print()).*/-comments with search and replace.

File: sechub-shared-kernel/src/main/java/com/daimler/sechub/sharedkernel/usecases/UseCaseIdentifier.java [hunk 6]
```diff
@@ -132,6 +132,8 @@
 
     UC_ADMIN_UNASSIGNS_EXECUTION_PROFILE_FROM_PROJECT(58),
 
+    /* project administration */
+    UC_UPDATE_PROJECT_METADATA(59),
```
[the remark this answers is not in the document]
Comment (de-jcup) [reply sechub-shared-kernel/src/main/java/com/daimler/sechub/sharedkernel/usecases/UseCaseIdentifier.java:136]: Discussed in Team: we make the rename

File: sechub-administration/src/test/java/com/daimler/sechub/domain/administration/project/ProjectUpdateAdministrationRestControllerMockTest.java (hunk 1 above)
Comment (4w31ss3) [reply sechub-administration/src/test/java/com/daimler/sechub/domain/administration/project/ProjectUpdateAdministrationRestControllerMockTest.java:124]: removed

File: sechub-doc/src/test/java/com/daimler/sechub/restdoc/ProjectUpdateAdministrationRestControllerRestDocTest.java (hunk 2 above)
Comment (4w31ss3) [reply sechub-doc/src/test/java/com/daimler/sechub/restdoc/ProjectUpdateAdministrationRestControllerRestDocTest.java:113]: removed

File: sechub-doc/src/test/java/com/daimler/sechub/restdoc/ProjectUpdateAdministrationRestControllerRestDocTest.java (hunk 3 above)
Comment (4w31ss3) [reply sechub-doc/src/test/java/com/daimler/sechub/restdoc/ProjectUpdateAdministrationRestControllerRestDocTest.java:116]: done

File: sechub-administration/src/main/java/com/daimler/sechub/domain/administration/project/ProjectMetaData.java (hunk 4 above)
Comment (4w31ss3) [reply sechub-administration/src/main/java/com/daimler/sechub/domain/administration/project/ProjectMetaData.java:12]: done

File: sechub-administration/src/main/java/com/daimler/sechub/domain/administration/project/ProjectMetaDataEntry.java (hunk 5 above)
Comment (4w31ss3) [reply sechub-administration/src/main/java/com/daimler/sechub/domain/administration/project/ProjectMetaDataEntry.java:31]: done

File: sechub-shared-kernel/src/main/java/com/daimler/sechub/sharedkernel/usecases/UseCaseIdentifier.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (4w31ss3) [reply sechub-shared-kernel/src/main/java/com/daimler/sechub/sharedkernel/usecases/UseCaseIdentifier.java:136]: done

File: sechub-administration/src/main/java/com/daimler/sechub/domain/administration/project/ProjectMetaDataEntry.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+import javax.persistence.Table;
+import javax.persistence.Version;
+import javax.persistence.Id;
+import javax.persistence.IdClass;
+import javax.persistence.JoinColumn;
+import javax.persistence.ManyToOne;
+import javax.persistence.Column;
+
+@Entity
+@IdClass(ProjectMetaData.class)
```
[the remark this answers is not in the document]
Comment (4w31ss3) [reply sechub-administration/src/main/java/com/daimler/sechub/domain/administration/project/ProjectMetaDataEntry.java]: done

File: sechub-administration/src/main/java/com/daimler/sechub/domain/administration/project/ProjectMetaDataRepository.java [hunk 8]
```diff
@@ -0,0 +1,7 @@
+package com.daimler.sechub.domain.administration.project;
+
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface ProjectMetaDataRepository extends JpaRepository<ProjectMetaDataEntry, ProjectMetaData> {
```
[the remark this answers is not in the document]
Comment (4w31ss3) [reply sechub-administration/src/main/java/com/daimler/sechub/domain/administration/project/ProjectMetaDataRepository.java]: done

File: sechub-administration/src/main/java/com/daimler/sechub/domain/administration/project/ProjectUpdateMetaDataService.java [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+		auditLog.log("triggers update of metadata for project {}. Updated metadata shall be {}", logSanitizer.sanitize(projectId, 30), metaData);
+
+		assertion.isValidProjectId(projectId);
+
+		Optional<Project> found = repository.findById(projectId);
+		if (!found.isPresent()) {
+			throw new NotFoundException("Project '" + projectId + "' does not exist or you have now ");
+		}
+		/*
+		 * TODO Albert Tregnaghi, 2018-09-06: currently we check only role SUPER_ADMIN.
```
[the remark this answers is not in the document]
Comment (4w31ss3) [reply sechub-administration/src/main/java/com/daimler/sechub/domain/administration/project/ProjectUpdateMetaDataService.java]: done

File: sechub-administration/src/main/java/com/daimler/sechub/domain/administration/project/ProjectUpdateMetaDataService.java [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+		 * Because super admin is the only role having access. But when we got a project
+		 * owner, the access to this project must be checked too! Here we should use
+		 * permissions instead of roles then
+		 */
+		Project project = found.get();
+		
+		// update is currently a replace action
+		metaDataRepository.deleteAll(project.getMetaData());
+		
+		List<ProjectMetaDataEntry> metaDataEntries = metaData.stream().map(entry -> new ProjectMetaDataEntry(projectId, entry.getKey(), entry.getValue())).collect(Collectors.toList());
```
[the remark this answers is not in the document]
Comment (4w31ss3) [reply sechub-administration/src/main/java/com/daimler/sechub/domain/administration/project/ProjectUpdateMetaDataService.java]: done

File: sechub-developertools/src/main/java/com/daimler/sechub/developertools/admin/ui/action/project/CreateProjectAction.java [hunk 11]
```diff
@@ -55,14 +57,31 @@ public void execute(ActionEvent e) {
 				whiteListURLs.add(uri.getValue());
```
Comment (4w31ss3) [inline sechub-developertools/src/main/java/com/daimler/sechub/developertools/admin/ui/action/project/CreateProjectAction.java:57]: Remove emptyString "" uris in here.

Comment (4w31ss3) [reply sechub-developertools/src/main/java/com/daimler/sechub/developertools/admin/ui/action/project/CreateProjectAction.java:57]: Don't forget emptyString handling in  UpdateProjectWhiteListAction

Comment (4w31ss3) [reply sechub-developertools/src/main/java/com/daimler/sechub/developertools/admin/ui/action/project/CreateProjectAction.java:57]: done

File: sechub-administration/src/main/java/com/daimler/sechub/domain/administration/project/ProjectCreationService.java [hunk 12]
```diff
@@ -67,7 +68,7 @@
 			name = "Create project",
 			description = "The service will create the project when not already existing with such name."))
 	/* @formatter:on */
-	public void createProject(@NotNull String projectId, @NotNull String description, @NotNull String owner, @NotNull Set<URI> whitelist) {
+	public void createProject(@NotNull String projectId, @NotNull String description, @NotNull String owner, @NotNull Set<URI> whitelist, @NotNull List<ProjectMetaData> metaData) {
```
[the remark this answers is not in the document]
Comment (4w31ss3) [reply sechub-administration/src/main/java/com/daimler/sechub/domain/administration/project/ProjectCreationService.java]: done

File: sechub-administration/src/main/java/com/daimler/sechub/domain/administration/project/ProjectAdministrationRestController.java [hunk 13]
```diff
@@ -2,6 +2,7 @@
 package com.daimler.sechub.domain.administration.project;
```
[the remark this answers is not in the document]
Comment (4w31ss3) [reply sechub-administration/src/main/java/com/daimler/sechub/domain/administration/project/ProjectAdministrationRestController.java:2]: done

**Review by de-jcup — APPROVED** [review]
Review has been done, PR will be merged. 

@4w31ss3 - :+1:  good work and thank you very much for the cleanup, renaming etc. done additionally inside this issue.

---
<sup>Albert Tregnaghi <albert.tregnaghi@daimler.com>, Daimler TSS GmbH, [imprint](https://github.com/Daimler/daimler-foss/blob/master/LEGAL_IMPRINT.md)</sup>

**Pull request merged**